### PR TITLE
Refactor `Expm` and add jax dispatch

### LINKED ISSUE #418: ENH: Add flag for `tensor.nlinalg.eig` to return complex part of eigenvalues
### Before

```python
# Currently the complex part of eigenvalues is discarded by `tensor.linalg.eig`:

x = np.array([[-0.2470853 , -0.92681485],
              [ 0.85821264,  1.24789551]])
v, w = pt.linalg.eig(x)
w.eval()
 
#  ComplexWarning: Casting complex values to real discards the imaginary part
#  w[0], v[0] = (z.astype(x.dtype) for z in self._numop(x))
```

### After

```python
v, (w_real, w_imag) = pt.linalg.eig(x, return_complex_part=True)
```

### Context for the issue:

I'm working on some problems that require analysis of modulus of eigenvalues, which I can't do now since the imaginary component is discarded. I know it's not possible to work directly with imaginary numbers in `pytensor`, but it would be easy enough to pass back the imaginary components to users in this case, who could then use them. 

The default would remain unchanged, to prevent legacy code from breaking

### PR DESCRIPTION
**Pull request by jessegrabowski**
## Description
Cleans up the `Expm` `Op` to be more modern. The gradient is now fully symbolic, and the separate gradient `Op` has been removed. In addition, the tests have been cleaned up and now use a parameterization.

Also add a `return_components` flag to the `Eig` op, which if `True` returns a stacked array of real and imaginary components of the eigenvalues. 

I included the change to `Eig` here  because the symbolic gradients of `Expm` depend on the eigenvalues of the input matrix, which in general will not be real. Truncating the imaginary components (which is our default) results in incorrect gradients. I imagine this is why the gradient computation was originally being done in a perform method.

## Related Issue

- [ ] Closes #418
- [ ] Related to #

## Checklist

- [ ] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [ ] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

----
📚 Documentation preview 📚: https://pytensor--1668.org.readthedocs.build/en/1668/

### CODE REVIEW

File: pytensor/tensor/slinalg.py [hunk 1]
```diff
@@ -1304,82 +1303,63 @@ def eigvalsh(a, b, lower=True):
...
 
-    """
+        exp_w = pt.exp(w)
+        numer = pt.sub.outer(exp_w, exp_w)
+        denom = pt.sub.outer(w, w)
 
-    __props__ = ()
+        # When w_i ≈ w_j, we have a removable singularity in the expression for X, because
+        # lim b->a (e^a - e^b) / (a - b) = e^a (derivation left for the motivated reader)
+        X = pt.where(pt.abs(denom) < 1e-8, exp_w, numer / denom)
```
[the remark this answers is not in the document]
Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py:1340]: wrong again, robot!

File: pytensor/tensor/nlinalg.py [hunk 2]
```diff
@@ -297,31 +305,84 @@ def slogdet(x: TensorLike) -> tuple[ptb.TensorVariable, ptb.TensorVariable]:
...
+        if return_components:
+            signature = "(m,m)->(a,m),(a,m,m)"
+            spec = ("numpy.linalg.eig", 2, 3)
+
+        else:
+            signature = "(m,m)->(m),(m,m)"
+            spec = ("numpy.linalg.eig", 1, 2)
+
+        self.gufunc_signature = signature
+        self.gufunc_spec = spec
```
Comment (ricardoV94) [inline pytensor/tensor/nlinalg.py]: spec numbers are about the number of inputs.

In any case this isn't good because Blockwise tries to use the function in the spec instead of doing it's own vectorization, when available. But it can't work like this because the new flag is not a behavior blockwise would know how to emulate. Simply take the `spec` thing out for now

File: pytensor/tensor/nlinalg.py [hunk 3]
```diff
@@ -335,6 +396,7 @@ class Eigh(Eig):
     def __init__(self, UPLO="L"):
         assert UPLO in ("L", "U")
         self.UPLO = UPLO
+        self.return_components = False
```
Comment (ricardoV94) [inline pytensor/tensor/nlinalg.py]: Commenting here because it's close. Is this another of those cases where Ops shouldn't be a subclass, but should share an abstract class? In rewrites tracking Eig, Eigh will also match

File: pytensor/tensor/slinalg.py [hunk 4]
```diff
@@ -1304,82 +1303,63 @@ def eigvalsh(a, b, lower=True):
...
-        # Kalbfleisch and Lawless, J. Am. Stat. Assoc. 80 (1985) Equation 3.4
-        # Kind of... You need to do some algebra from there to arrive at
-        # this expression.
-        (A, gA) = inputs
-        (out,) = outputs
-        w, V = scipy_linalg.eig(A, right=True)
-        U = scipy_linalg.inv(V).T
+        # At this point, result is always a complex dtype. If the input was real, the output should be
+        # real as well (and all the imaginary parts are numerical noise)
+        if A.dtype not in ("complex64", "complex128"):
```
Comment (ricardoV94) [inline pytensor/tensor/slinalg.py:1350]: raise Explicitly otherwise, or there will be a weird message from PyTensor saying an Op returned an invalid complex value (better for the Op to return NotImplemented, so it doesn't seem like a bug)?

Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py:1350]: The idea here was that A might have been complex to begin with

File: pytensor/tensor/nlinalg.py (hunk 3 above)
Comment (jessegrabowski) [reply pytensor/tensor/nlinalg.py]: Maybe, yeah. A subclass might be fine because they can share gradients? I haven't spend any time looking at eigh

File: pytensor/tensor/nlinalg.py (hunk 2 above)
Comment (jessegrabowski) [reply pytensor/tensor/nlinalg.py]: Ok. I just guessed on what the numbers meant

File: tests/tensor/test_slinalg.py [hunk 5]
```diff
@@ -880,35 +880,26 @@ def test_expm():
...
-    rng = np.random.default_rng(utt.fetch_seed())
-    # Always test in float64 for better numerical stability.
-    A = rng.standard_normal((5, 5))
-
-    utt.verify_grad(expm, [A], rng=rng)
+@pytest.mark.parametrize(
+    "mode", ["symmetric", "nonsymmetric_real_eig", "nonsymmetric_complex_eig"][-1:]
+)
+def test_expm_grad(mode):
+    rng = np.random.default_rng()
```
[the remark this answers is not in the document]
Comment (jessegrabowski) [reply tests/tensor/test_slinalg.py:887]: Simply not true

File: pytensor/tensor/nlinalg.py (hunk 2 above)
Comment (ricardoV94) [reply pytensor/tensor/nlinalg.py]: Or leave if you are not stacking, that's still nice to use numpy for handling the batch dims than having to blockwise and call the core_op needlessly

File: pytensor/tensor/slinalg.py (hunk 4 above)
Comment (ricardoV94) [reply pytensor/tensor/slinalg.py:1350]: Sure but we don't support complex gradients

File: pytensor/tensor/nlinalg.py [hunk 6]
```diff
@@ -297,31 +305,84 @@ def slogdet(x: TensorLike) -> tuple[ptb.TensorVariable, ptb.TensorVariable]:
...
-
     """
 
-    __props__: tuple[str, ...] = ()
-    gufunc_signature = "(m,m)->(m),(m,m)"
-    gufunc_spec = ("numpy.linalg.eig", 1, 2)
+    __props__: tuple[str, ...] = ("return_components",)
+
+    def __init__(self, return_components: bool = False):
+        self.return_components = return_components
```
Comment (ricardoV94) [inline pytensor/tensor/nlinalg.py]: what about the dispatches for other backends?

Comment (jessegrabowski) [reply pytensor/tensor/nlinalg.py]: Good question. The whole return_components thing has the feeling of a hack, do you have any ideas for alternatives? 

Jax allows the domain change:

```py
import numpy as np
import jax

x = np.random.normal(size=(5, 5))
w, v = jax.numpy.linalg.eig(x)
print(x.dtype, w.dtype, v.dtype)
# float64 complex64 complex64
```

Numba does not and errors out:

```py
import numba as nb
import numpy as np

@nb.njit
def nb_eig(x):
    return np.linalg.eig(x)

x = np.random.normal(size=(5, 5))
try:
    nb_eig(x)
except ValueError as e:
    print(e)
# eig() argument must not cause a domain change.
```

MLX allows the recast as well:

```
import mlx.core as mx
import numpy as np
x = mx.array(np.random.normal(size=(5, 5)))
w, v = mx.linalg.eig(x, stream=mx.cpu)
print(x.dtype, w.dtype, v.dtype)
# mlx.core.float32 mlx.core.complex64 mlx.core.complex64
```

So we're kind of the odd-man out. 

(In case you're curious about gradients, either jax nor mlx support gradients for `eig`, only for `eigh`)

File: pytensor/tensor/slinalg.py (hunk 4 above)
Comment (jessegrabowski) [reply pytensor/tensor/slinalg.py:1350]: I want to respect the user's dtype. If we get a complex input, the gradient computation will have bigger problems than just this one op.

File: pytensor/tensor/nlinalg.py (hunk 6 above)
Comment (ricardoV94) [reply pytensor/tensor/nlinalg.py]: I'm just talking about  the extra behavior of `stacked_components` that you added. That should be emulated in the dispatched functions even if not needed for their auto-diff. Or raise a NotImplementedError if you dont' want to support it.

Otherwise you promised the output would be 2d but it's actually 3d

File: pytensor/tensor/nlinalg.py [hunk 7]
```diff
@@ -297,31 +305,84 @@ def slogdet(x: TensorLike) -> tuple[ptb.TensorVariable, ptb.TensorVariable]:
...
+
+        if return_components:
+            signature = "(m,m)->(a,m),(a,m,m)"
```
Comment (ricardoV94) [inline pytensor/tensor/nlinalg.py]: Numpy signatures can have constants. We use this information as well in some cases

```suggestion
            signature = "(m,m)->(2,m),(2,m,m)"
```

File: pytensor/tensor/nlinalg.py [hunk 8]
```diff
@@ -297,31 +305,84 @@ def slogdet(x: TensorLike) -> tuple[ptb.TensorVariable, ptb.TensorVariable]:
...
+def eig(x: TensorLike, return_components: bool = False):
+    """
+    Return the eigenvalues and right eigenvectors of a square array.
+
+    Parameters
+    ----------
+    x: TensorLike
+        Square matrix, or array of such matrices
+    return_components: bool, optional
+        By default, only the real component of the eigenvalues and eigenvectors are returned, as PyTensor does not
```
Comment (ricardoV94) [inline pytensor/tensor/nlinalg.py]: Something more explicit, along the lines of:

Numpy eig can return either float or complex numbers depending on the runtime inputs. This is not supported by PyTensor, which requires a static dtype known in advance.

*** 

Alternatively we could always return complex numbers but it's going to break autodiff all the time?

File: pytensor/tensor/nlinalg.py (hunk 6 above)
Comment (jessegrabowski) [reply pytensor/tensor/nlinalg.py]: Yes I understand what you wanted, I was using this as a chance to talk about the approach in general

File: pytensor/tensor/nlinalg.py (hunk 8 above)
Comment (jessegrabowski) [reply pytensor/tensor/nlinalg.py]: > Alternatively we could always return complex numbers but it's going to break autodiff all the time?

That's basically what jax does. If the user knows his system has only positive eigenvalues (e.g. it's a PSD matrix), he can use `pt.linalg.eigh`. 

For the record, we don't have gradients for `Eig` anyway

Comment (ricardoV94) [reply pytensor/tensor/nlinalg.py]: So why not return complex all the time then?

Comment (jessegrabowski) [reply pytensor/tensor/nlinalg.py]: Honestly? Because I thought about that idea after I already opened this PR.

**Comment by jessegrabowski** [comment]
@ricardoV94 I updated the PR to have `Eig` always return a complex dtype, so we now match what JAX does. I put an informative error in the `L_op` method, telling the user to go use `eigh` if she wants gradients.

File: pytensor/tensor/nlinalg.py [hunk 9]
```diff
@@ -297,44 +305,86 @@ def slogdet(x: TensorLike) -> tuple[ptb.TensorVariable, ptb.TensorVariable]:
...
     Return the eigenvalues and eigenvectors of a Hermitian or symmetric matrix.
-
     """
 
     __props__ = ("UPLO",)
 
     def __init__(self, UPLO="L"):
         assert UPLO in ("L", "U")
         self.UPLO = UPLO
+        self.return_components = False
```
Comment (ricardoV94) [inline pytensor/tensor/nlinalg.py]: still here

**Comment by ricardoV94** [comment]
looks good just a leftover from the previous iteration
